_posts/2019-08-29-using-tf-from-r/using-tf-from-R.Rmd

```diff
@@ -0,0 +1,413 @@
+
+
+```{r}
+model <- keras_model_sequential()
+model$add(layer_dense(units = 32L))
+```
+
+we can just say
+
+```{r}
+model <- keras_model_sequential()
+model %>% layer_dense(units = 32)
+```
+
+While these are nice to have, there is more. Advanced functionality in (Python) Keras mostly depends on the ability to subclass objects. One example is custom callbacks. If you were using Python, you'd have to subclass `tf.keras.callbacks.Callback`. From R, you can create an R6 class inheriting from `KerasCallback`, like so
+
+```{r}
+CustomCallback <- R6::R6Class("CustomCallback",
+    inherit = KerasCallback,
+    public = list(
+      on_train_begin = function(logs) {
+        # do something
+      },
+      on_train_end = function(logs) {
+        # do something
+      }
+    )
+  )
+```
+
+This is because `keras` defines an actual Python class, `RCallback`, and maps your R6 class'es methods to it.
+Another example are [custom models](https://tensorflow.rstudio.com/keras/articles/custom_models.html), introduced on this blog [about a year ago](https://blogs.rstudio.com/tensorflow/posts/2018-07-30-attention-layer/).
+These models can be trained with custom training loops. In R, you use `keras_model_custom` to create one, for example, like this:
+
+```{r}
+m <- keras_model_custom(name = "mymodel", function(self) {
+  self$dense1 <- layer_dense(units = 32, activation = "relu")
+  self$dense2 <- layer_dense(units = 10, activation = "softmax")
+  
+  function(inputs, mask = NULL) {
+    self$dense1(inputs) %>%
+      self$dense2()
+  }
+})
+```
+
+Here, `keras` will make sure an actual Python object is created that subclasses `tf.keras.Model` and when called, runs the above anonymous `function()`.
+
+So that's `keras`. What about the `tensorflow` package? As a user you only need it when you have to do advanced stuff, like configure TensorFlow device usage or (in TF 1.x) access elements of the `Graph` or the `Session`. Internally, it is used by `keras` heavily. Essential internal functionality includes, e.g., implementations of S3 methods, like `print`, `[` or `+`, on `Tensor`s, so you can operate on them like on R vectors.  
+
+Now that we know what each of the packages is "for", let's dig deeper into what makes this possible.
+
+
+## Show me the magic: reticulate
+
+Instead of exposing the topic top-down, we follow a by-example approach, building up complexity as we go. We'll have three scenarios.
+
+First, we assume we already have a Python object (that has been constructed in whatever way) and need to convert that to R. Then, we'll investigate how we can create a Python object, calling its constructor. Finally, we go the other way round: We ask how we can pass an R function to Python for later usage.
+
+### Scenario 1: R-to-Python conversion
```

Comment:
yeah, if I say it at that point I basically give up the plot ... Right now I'm not sure if the plot doesn't work so well, or if you see it like that because you're reading it as _the expert_ ...
Maybe I leave this open and make the other changes, and when you see it again, let me know if you still think like on first read?

Reply:
To add to this, I framed this as an exploration / reverse engineering task because for me it basically was. So as long as "the story works" && everything is correct I think the bottom-up approach  should be ok - for now - but that by no means precludes another, later, post which would be written as a top-down exposition ... (actually I'd like to write such a post, at a later time)